src/System.Private.CoreLib/src/System/Threading/Timer.CoreCLR.cs

```diff
@@ -0,0 +1,106 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using Microsoft.Win32;
+using Microsoft.Win32.SafeHandles;
+using System.Diagnostics;
+using System.Runtime.CompilerServices;
+using System.Runtime.InteropServices;
+
+namespace System.Threading
+{
+    internal partial class TimerQueue
+    {
+        #region interface to native per-AppDomain timer
+
+        private static int TickCount
+        {
+            get
+            {
+#if !FEATURE_PAL
+                // We need to keep our notion of time synchronized with the calls to SleepEx that drive
+                // the underlying native timer.  In Win8, SleepEx does not count the time the machine spends
+                // sleeping/hibernating.  Environment.TickCount (GetTickCount) *does* count that time,
+                // so we will get out of sync with SleepEx if we use that method.
+                //
+                // So, on Win8, we use QueryUnbiasedInterruptTime instead; this does not count time spent
+                // in sleep/hibernate mode.
+                if (Environment.IsWindows8OrAbove)
+                {
+                    ulong time100ns;
+
+                    bool result = Win32Native.QueryUnbiasedInterruptTime(out time100ns);
+                    if (!result)
+                        throw Marshal.GetExceptionForHR(Marshal.GetLastWin32Error());
+
+                    // convert to 100ns to milliseconds, and truncate to 32 bits.
+                    return (int)(uint)(time100ns / 10000);
+                }
+                else
+#endif
+                {
+                    return Environment.TickCount;
+                }
+            }
+        }
+
+        // We use a SafeHandle to ensure that the native timer is destroyed when the AppDomain is unloaded.
+        private sealed class AppDomainTimerSafeHandle : SafeHandleZeroOrMinusOneIsInvalid
+        {
+            public AppDomainTimerSafeHandle()
+                : base(true)
+            {
+            }
+
+            protected override bool ReleaseHandle()
+            {
+                return DeleteAppDomainTimer(handle);
+            }
+        }
+
+        private readonly int _id; // TimerQueues[_id] == this
```

Comment:
I have not realized that the `_id` is an index into TimerQueues array that is in the shared partition. In that case, the  `private readonly int _id;` and `private TimerQueue(int id)` would be fine in the shared partition.

Reply:
OTOH it is used only by two out of the three implementations, so both approaches make some sense. I can move it back though.